actors/market/src/lib.rs

```diff
@@ -537,139 +545,228 @@ impl Actor {
+            Ok((activations, batch_gen.gen()))
+        })?;
 
-                        if s.is_some() {
-                            return Err(actor_error!(
-                                illegal_argument,
-                                "deal {} already activated",
-                                deal_id
-                            ));
-                        }
+        Ok(BatchActivateDealsResult { activations, activation_results: batch_ret })
+    }
 
-                        let propc = rt_deal_cid(rt, proposal)?;
+    /// Receives notification of a change to sector content, which may satisfy to activate a deal.
+    /// Deals are activated or fail independently, including in the same sector.
+    /// This is an alternative to ActivateDeals.
+    fn sector_content_changed(
+        rt: &impl Runtime,
+        params: ext::miner::SectorContentChangedParams,
+    ) -> Result<ext::miner::SectorContentChangedReturn, ActorError> {
+        rt.validate_immediate_caller_type(std::iter::once(&Type::Miner))?;
+        let miner_addr = rt.message().caller();
+        let curr_epoch = rt.curr_epoch();
 
-                        // Confirm the deal is in the pending proposals queue.
-                        // It will be removed from this queue later, during cron.
-                        let has = st.has_pending_deal(rt.store(), propc)?;
+        let sectors_ret = rt.transaction(|st: &mut State, rt| {
+            let proposals = st.load_proposals(rt.store())?;
+            let states = st.load_deal_states(rt.store())?;
+            let pending_deals = st.load_pending_deals(rt.store())?;
+            let mut pending_deal_allocation_ids =
+                st.load_pending_deal_allocation_ids(rt.store())?;
 
-                        if !has {
-                            return Err(actor_error!(
-                                illegal_state,
-                                "tried to activate deal that was not in the pending set ({})",
-                                propc
-                            ));
+            let mut deal_states: Vec<(DealID, DealState)> = vec![];
+            let mut activated_deals: HashSet<DealID> = HashSet::new();
+            let mut sectors_deals: Vec<(SectorNumber, SectorDealIDs)> = vec![];
+            let mut sectors_ret: Vec<ext::miner::SectorReturn> = vec![];
+
+            for sector in &params.sectors {
+                let mut sector_deal_ids: Vec<DealID> = vec![];
+                let mut pieces_ret: Vec<ext::miner::PieceReturn> = vec![];
+                for piece in &sector.added {
+                    let deal_id: DealID = match deserialize(&piece.payload.clone(), "deal id") {
+                        Ok(v) => v,
+                        Err(e) => {
+                            log::warn!("failed to deserialize deal id {:?}: {}", piece.payload, e);
+                            pieces_ret.push(ext::miner::PieceReturn { accepted: false });
```

Comment:
So, code like this always concerns me a bit as its easy to forget to push in some error case.

How about:

1. Pre-filling the `pieces_ret` with `vec![PieceReturn { accepted: false }; sector.added.len()]`.
2. Iterating over `for (ret, piece) in sectors.added.iter().zip(&mut pieces_ret) { ...`
3. Setting `*ret.accepted = true` at the end?

Or you can use map/collect.

Reply:
Nice, will do